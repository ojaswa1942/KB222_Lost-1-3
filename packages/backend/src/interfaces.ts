export interface JWTPayload {
  email: string;
  id: string;
}
