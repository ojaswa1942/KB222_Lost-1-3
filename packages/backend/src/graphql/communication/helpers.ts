import { getRepository } from 'typeorm';
import { Room } from '../../database/entity/Room';
import { Channel } from '../../database/entity/Channel';
import { User } from '../../database/entity/User';

export const createRoom = async (name: string, channel: Channel, users: User[]): Promise<Room> => {
  const roomRepo = getRepository(Room);
  const newRoom = roomRepo.create({ name, channel, users });

  return roomRepo.save(newRoom);
};
