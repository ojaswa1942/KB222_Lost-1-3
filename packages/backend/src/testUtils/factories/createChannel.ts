import { Channel } from '../../database/entity/Channel';
import { createEntity } from './createEntity';

export async function createChannel(attributes: Partial<Channel> = {}): Promise<Channel> {
  return createEntity(Channel, {
    ...attributes,
  });
}
