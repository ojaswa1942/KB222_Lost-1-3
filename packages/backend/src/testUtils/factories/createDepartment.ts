import faker from 'faker';

import { Department } from '../../database/entity';
import { createEntity } from './createEntity';

export async function createDepartment(attributes: Partial<Department> = {}): Promise<Department> {
  return createEntity(Department, {
    name: `${faker.name.jobArea()} Department`,
    ...attributes,
  });
}
