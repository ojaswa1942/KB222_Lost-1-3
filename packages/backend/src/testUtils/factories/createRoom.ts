import faker from 'faker';

import { Room } from '../../database/entity/Room';
import { createEntity } from './createEntity';

export async function createRoom(attributes: Partial<Room> = {}): Promise<Room> {
  return createEntity(Room, {
    name: `${faker.address.city()}`,
    ...attributes,
  });
}
