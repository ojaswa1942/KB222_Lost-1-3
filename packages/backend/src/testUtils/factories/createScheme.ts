import faker from 'faker';

import { Scheme } from '../../database/entity';
import { createEntity } from './createEntity';

export async function createScheme(attributes: Partial<Scheme> = {}): Promise<Scheme> {
  return createEntity(Scheme, {
    name: `${faker.name.jobArea()} Scheme`,
    ...attributes,
  });
}
