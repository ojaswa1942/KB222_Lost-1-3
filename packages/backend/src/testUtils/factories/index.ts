export { createUser } from './createUser';
export { createDepartment } from './createDepartment';
export { createScheme } from './createScheme';
export { createChannel } from './createChannel';
export { createRoom } from './createRoom';
