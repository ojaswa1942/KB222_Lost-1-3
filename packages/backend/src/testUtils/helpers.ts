import { ApolloServer } from 'apollo-server-express';
import { ApolloServerTestClient, createTestClient as createApolloTestClient } from 'apollo-server-testing';

import { Context } from '../context';
import { rootSchema } from '../graphql/rootSchema';
import {
  buildUserLoader,
  buildDepartmentLoader,
  buildSchemeLoader,
  buildChannelLoader,
  buildRoomLoader,
  buildMessageLoader,
} from '../loaders';

export const createTestClient = (ctx?: Partial<Context>): ApolloServerTestClient => {
  const server = new ApolloServer({
    schema: rootSchema,
    context: {
      headers: {},
      req: {},
      res: {
        cookie: () => {
          // do nothing
        },
      },
      userLoader: buildUserLoader(),
      departmentLoader: buildDepartmentLoader(),
      schemeLoader: buildSchemeLoader(),
      channelLoader: buildChannelLoader(),
      roomLoader: buildRoomLoader(),
      messageLoader: buildMessageLoader(),
      isValid: false,
      ...ctx,
    },
  });

  return createApolloTestClient(server);
};
