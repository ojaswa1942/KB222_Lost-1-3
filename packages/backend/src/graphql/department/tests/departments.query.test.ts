import { gql } from 'apollo-server-express';

import { createDepartment, createUser } from '../../../testUtils/factories';
import { createTestClient } from '../../../testUtils/helpers';
import { UserType } from '../../../interfaces';

describe('departments query', () => {
  const DepartmentsQuery = gql`
    query {
      departments {
        id
      }
    }
  `;

  test('fetch departments with root', async () => {
    // Given
    const user = await createUser({ email: 'root@example.com', password: 'test', type: UserType.Root });
    const d1 = createDepartment();
    const d2 = createDepartment();
    const d3 = createDepartment();
    await Promise.all([d1, d2, d3]);

    // When
    const res = await createTestClient({
      isValid: true,
      jwt: { id: user.id, email: user.email, type: user.type },
    }).query({
      query: DepartmentsQuery,
    });

    // Then
    expect(res.errors).toBe(undefined);
    expect(res.data).not.toBe(null);
    expect(res.data?.departments).toMatchSnapshot();
  });

  test('fetch departments with state user', async () => {
    // Given
    const user = await createUser({ email: 'root@example.com', password: 'test', type: UserType.State });
    await createDepartment();
    // await createDepartment({ users: [user] });
    // await createDepartment();
    // await createDepartment({ users: [user] });

    // When
    const res = await createTestClient({
      isValid: true,
      jwt: { id: user.id, email: user.email, type: user.type },
    }).query({
      query: DepartmentsQuery,
    });

    // Then
    expect(res.errors).toBe(undefined);
    expect(res.data).not.toBe(null);
    expect(res.data?.departments).toMatchSnapshot();
  });
});
