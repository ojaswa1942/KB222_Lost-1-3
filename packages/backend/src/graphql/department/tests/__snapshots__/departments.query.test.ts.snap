// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`departments query fetch departments with root 1`] = `
Array [
  Object {
    "id": 1,
    "name": "Solutions Department",
  },
  Object {
    "id": 2,
    "name": "Group Department",
  },
  Object {
    "id": 3,
    "name": "Functionality Department",
  },
]
`;

exports[`departments query fetch departments with state user 1`] = `
Array [
  Object {
    "id": 2,
    "name": "Functionality Department",
  },
  Object {
    "id": 4,
    "name": "Functionality Department",
  },
]
`;
